components/external_storage/src/local.rs

```diff
@@ -16,44 +16,31 @@ use rand::Rng;
 
 use super::{util::error_stream, ExternalStorage};
 
-const LOCAL_STORAGE_TMP_DIR: &str = "localtmp";
 const LOCAL_STORAGE_TMP_FILE_SUFFIX: &str = "tmp";
 
-fn maybe_create_dir(path: &Path) -> io::Result<()> {
-    if let Err(e) = fs::create_dir_all(path) {
-        if e.kind() != io::ErrorKind::AlreadyExists {
-            return Err(e);
-        }
-    }
-    Ok(())
-}
-
 /// A storage saves files in local file system.
 #[derive(Clone)]
 pub struct LocalStorage {
     base: PathBuf,
     base_dir: Arc<File>,
-    tmp: PathBuf,
 }
 
 impl LocalStorage {
     /// Create a new local storage in the given path.
     pub fn new(base: &Path) -> io::Result<LocalStorage> {
         info!("create local storage"; "base" => base.display());
-        let tmp_dir = base.join(LOCAL_STORAGE_TMP_DIR);
-        maybe_create_dir(&tmp_dir)?;
         let base_dir = Arc::new(File::open(base)?);
         Ok(LocalStorage {
             base: base.to_owned(),
             base_dir,
-            tmp: tmp_dir,
         })
     }
 
     fn tmp_path(&self, path: &Path) -> PathBuf {
         let uid: u64 = rand::thread_rng().gen();
         let tmp_suffix = format!("{}{:016x}", LOCAL_STORAGE_TMP_FILE_SUFFIX, uid);
-        self.tmp.join(path).with_extension(tmp_suffix)
+        // Save tmp files in base directory.
+        self.base.join(path).with_extension(tmp_suffix)
```

Comment:
would this litter the target storage with `*.tmpXXXXXXXXXXXXXXXX` files in some rare case of failure between `File::create` to `fs::rename` 😅

Reply:
Right, compare with possible data corruption (overwrite or partial write), I think it's better to leave unnecessary litter 😅